### PR TITLE
Support RouteUrlObject as redirect.destination in GetServerSideProps/GetStaticProps

### PR DESCRIPTION
**Pull request by s-r-x**
basically it just converts RouteUrlObject to string, using resolveHref function from next/router module. resolveHref requires next router, but it only uses pathname and aspath props from one. so i created a server version, which mocks router as  {pathname: '/', aspath: '/'}. this version always passes resolveAs: true to resolveHref, in order to normalize routes like /blog/post/[id]?id=1 to /blog/post/1.
i ran the auth example from examples folder with new updates, and it seems to work fine, although some tests are definitely required

## Bug Checklist

- [ ] Integration test added (see [test docs](https://blitzjs.com/docs/contributing#running-tests) if needed)

## Feature Checklist

- [ ] Integration test added (see [test docs](https://blitzjs.com/docs/contributing#running-tests) if needed)
- [ ] Documentation added/updated (submit PR to [blitzjs.com repo `canary` branch](https://github.com/blitz-js/blitzjs.com/tree/canary))

### CODE REVIEW

**Comment by beerose** [comment]
Does this close this issue: https://github.com/blitz-js/blitz/issues/2519?

**Comment by s-r-x** [comment]
@beerose oh sorry, i forgot to link it. yes, that's the issue this pr is aiming for.

File: nextjs/packages/next/shared/lib/router/router.ts [hunk 1]
```diff
@@ -269,12 +269,17 @@ function omitParmsFromQuery(query: ParsedUrlQuery, params: string[]) {
...
+  return resolved[1] || resolved[0]
+}
+
 /**
  * Resolves a given hyperlink with a certain router state (basePath not included).
  * Preserves absolute urls.
  */
 export function resolveHref(
-  router: NextRouter,
+  router: Pick<NextRouter, 'asPath' | 'pathname'>,
```
Comment (beerose) [inline nextjs/packages/next/shared/lib/router/router.ts:282]: Why was this change needed?

File: nextjs/packages/next/server/render.tsx [hunk 2]
```diff
@@ -381,6 +382,14 @@ function checkRedirectValues(
   }
 }
 
+function normalizeRedirectValues(redirect: Redirect) {
+  const dest = redirect.destination
+  if (dest && typeof dest === 'object') {
+    const resolvedDest = resolveHrefOnServer(dest)
```
Comment (beerose) [inline nextjs/packages/next/server/render.tsx:388]: We can re-use `formatWithValidation` function here — https://github.com/blitz-js/blitz/blob/canary/nextjs/packages/next/shared/lib/utils.ts#L567
```suggestion
    const resolvedDest = formatWithValidation(dest)
```

File: nextjs/packages/next/shared/lib/router/router.ts [hunk 3]
```diff
@@ -269,12 +269,17 @@ function omitParmsFromQuery(query: ParsedUrlQuery, params: string[]) {
   return filteredQuery
 }
 
+export function resolveHrefOnServer(href: Url): string {
```
Comment (beerose) [inline nextjs/packages/next/shared/lib/router/router.ts:272]: Changes in this file can be reverted if we use `formatWithValidation` in `normalizeRedirectValues` function

**Review by beerose — CHANGES_REQUESTED** [review]
Thanks for this PR! 🙏 And sorry for the delay with the review!

File: nextjs/packages/next/server/render.tsx (hunk 2 above)
Comment (s-r-x) [reply nextjs/packages/next/server/render.tsx:388]: actually, using formatWithValidation was my first idea. it works, but with a catch! i'll capture a video later to demonstrate you

Comment (s-r-x) [reply nextjs/packages/next/server/render.tsx:388]: with formatWithValidation:
https://user-images.githubusercontent.com/41614937/142560305-21131556-e348-4430-8fc0-1ab1c76967f6.mp4

with resolveHrefOnServer(i don't really like the name though):
https://user-images.githubusercontent.com/41614937/142560313-589eafcf-1977-4d0e-89a4-c1859b59eebb.mp4

Comment (beerose) [reply nextjs/packages/next/server/render.tsx:388]: Okay, got it. Thanks for the recordings! 

@flybayer It looks like we actually need to use `resolveHref` function as we're probably interested in the `interpolatedAs` value: https://github.com/blitz-js/blitz/blob/canary/nextjs/packages/next/shared/lib/router/router.ts#L336

Comment (flybayer) [reply nextjs/packages/next/server/render.tsx:388]: @s-r-x ok makes sense. Is there any issue with hardcoding `{ pathname: '/', asPath: '/' }`? Is there any edge cases this could cause a problem for? It would be good to add a clear comment explaining this for future folks.

Comment (s-r-x) [reply nextjs/packages/next/server/render.tsx]: @flybayer, @beerose  i replaced that hacky router mock with real server router(basically i just remove that resolveHrefOnServer util and revert the type of resolveHref router from { pathname: string, asPath: string } back to NextRouter). final result is basically the same, but the solution is less hacky. and i think there is no need for additional comments anymore.

**Comment by beerose** [comment]
Could you also add a mention of that to the docs? (to the `getStaticProps` and `getServerSideProps` sections)

**Review by beerose — COMMENTED** [review]
~It looks like Nexjs test #1 fails consistently (`server-side only compilation errors`). It works locally though~
